services/121-service/src/transaction-job-processors/transaction-job-processors.service.ts

```diff
@@ -224,18 +228,119 @@ export class TransactionJobProcessorsService {
+
+    // 1. Get registration details needed
+    // TODO: this is duplicate code with Visa-method > simplify
+    const registration =
+      await this.registrationScopedRepository.getByReferenceId({
+        referenceId: input.referenceId,
+      });
+    if (!registration) {
+      throw new Error(
+        `Registration was not found for referenceId ${input.referenceId}`,
+      );
+    }
+    const oldRegistration = structuredClone(registration);
+    const financialServiceProvider =
+      await this.financialServiceProviderRepository.getByName(
+        FinancialServiceProviderName.safaricom,
+      );
+    if (!financialServiceProvider) {
+      throw new Error('Financial Service Provider not found');
+    }
+
+    // 2. Check if all required properties are present. If not, create a failed transaction and throw an error.
+    for (const [name, value] of Object.entries(input)) {
+      // TODO: make some properties optional like in Visa, but why?
+
+      // Define "empty" based on your needs. Here, we check for null, undefined, or an empty string.
+      if (value === null || value === undefined || value === '') {
+        const errorText = `Property ${name} is undefined`;
+        await this.createTransactionAndUpdateRegistration({
+          programId: input.programId,
+          paymentNumber: input.paymentNumber,
+          userId: input.userId,
+          calculatedTransferAmountInMajorUnit: input.transactionAmount,
+          financialServiceProviderId: financialServiceProvider.id,
+          registration,
+          oldRegistration,
+          isRetry: input.isRetry,
+          status: StatusEnum.error,
+          errorText,
+        });
+        return;
+      }
+    }
+
+    // 3. Start the transfer, save error transaction on failure
+    let safaricomDoTransferResult: DoTransferReturnParams;
+    try {
+      safaricomDoTransferResult = await this.safaricomService.doTransfer({
+        transactionAmount: input.transactionAmount,
+        programId: input.programId,
+        paymentNr: input.paymentNumber,
+        userId: input.userId,
+        referenceId: input.referenceId,
+        registrationProgramId: input.registrationProgramId,
+        phoneNumber: input.phoneNumber,
+        nationalId: input.nationalId,
+      });
+    } catch (error) {
+      await this.createTransactionAndUpdateRegistration({
+        programId: input.programId,
+        paymentNumber: input.paymentNumber,
+        userId: input.userId,
+        calculatedTransferAmountInMajorUnit: input.transactionAmount,
+        financialServiceProviderId: financialServiceProvider.id,
+        registration,
+        oldRegistration,
+        isRetry: input.isRetry,
+        status: StatusEnum.error,
+        errorText: error?.message,
+      });
+      return;
+    }
+
+    // 4. If transfer is successful, create message and add to queue (not needed for safaricom)
```

Comment:
If transfer was not successful: update the created transaction from success to error.